### PR TITLE
Update ConstantTimeUtility.cs

### PR DESCRIPTION
**Pull request by kov072299**
- Fix typo

# Description

Typo in ConstantTimeEq meant method only worked as intended for inputs qword aligned (8 byte-aligned). 

## Type of change

- [x] Bug fix (non-breaking change which fixes an issue)

# Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes
- [ ] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

**Comment by cschuchardt88** [comment]
Can you add a test?

**Comment by diskooooo** [comment]
Unknown contributor, new to Github this month, without introduction, sends in a PR with a change to a NEO cryptography library... things that make me go "hmmmm" 🤔

**Comment by vncoelho** [comment]
@txhsl, what do you think?
@doubiliu

**Review by vncoelho — APPROVED** [review]
I believe that the change is correct because this line is intended to compare corresponding bytes from `a_bytes` and `b_bytes` and accumulate any differences into the `f` variable. The use of `XOR (^)` will result in a non-zero value for any byte that differs between `a_bytes` and `b_bytes`, and the `f` variable accumulates these differences. If `f` ends up being `0`, it means the two byte arrays are equal.

In this sense, without this change we just compare each byte of `a_bytes` with itself, which would always result in 0, effectively rendering the check meaningless and always producing a result of true for equality, which looks like to be incorrect.

Please double check before merging.

**Review by vncoelho — COMMENTED** [review]
During test it would be good to ensure that the `ConstantTimeEq` method behaves as expected in all cases, particularly for various data types.

If we can design some UTs based on that it would be great.

**Review by roman-khimov — COMMENTED** [review]
But why do we have `ConstantTimeEq` at all in our code? Isn't something like this provided by standard libraries? Like https://learn.microsoft.com/en-us/dotnet/api/system.security.cryptography.cryptographicoperations.fixedtimeequals?view=net-8.0?

**Review by shargon — APPROVED** [review]

**Comment by Jim8y** [comment]
> But why do we have `ConstantTimeEq` at all in our code? Isn't something like this provided by standard libraries? Like https://learn.microsoft.com/en-us/dotnet/api/system.security.cryptography.cryptographicoperations.fixedtimeequals?view=net-8.0?

it was converted from rust lib from file to file, but i agree with you, @cschuchardt88 @vncoelho @shargon do you think we should adopt dotnet lib method?

**Comment by shargon** [comment]
Let's fix this, and create an issue for the native one

File: src/Neo.Cryptography.BLS12_381/ConstantTimeUtility.cs [hunk 1]
```diff
@@ -26,7 +26,7 @@ public static bool ConstantTimeEq<T>(in T a, in T b) where T : unmanaged
         for (int i = 0; i < a_u64.Length; i++)
             f |= a_u64[i] ^ b_u64[i];
         for (int i = a_u64.Length * sizeof(ulong); i < a_bytes.Length; i++)
-            f |= (ulong)a_bytes[i] ^ a_bytes[i];
+            f |= (ulong)a_bytes[i] ^ b_bytes[i];
```
Comment (AnnaShaleva) [inline src/Neo.Cryptography.BLS12_381/ConstantTimeUtility.cs:29]: Quite a straightforward fix, just a question from my side: is it possible that `b_bytes` has the length less than `a_bytes`? We're iterating up to `a_bytes.Length`, so this code is valid if these spans have the same length.

**Comment by txhsl** [comment]
> @txhsl, what do you think?

At the beginning, I was thinking about how things like this passed tests, so I checked the context. :)

So first, it is not a transfer from Rust. If we are talking about https://github.com/zkcrypto/bls12_381, the codes are different.

Second, this method is checking two same-type inputs, so it skips the length check.

Third, is this line really used? If `a_u64.Length * sizeof(ulong)` always equals `a_bytes.Length`, then that line is never reached.

https://github.com/neo-project/neo/blob/90fbbe1d0c1ededdd8e0c3048a63878b4b319bc3/src/Neo.Cryptography.BLS12_381/ConstantTimeUtility.cs#L28-L29

File: src/Neo.Cryptography.BLS12_381/ConstantTimeUtility.cs (hunk 1 above)
Comment (shargon) [reply src/Neo.Cryptography.BLS12_381/ConstantTimeUtility.cs:29]: Yes we need to check the length before

**Comment by AnnaShaleva** [comment]
> If a_u64.Length * sizeof(ulong) always equals a_bytes.Length

It's true in case if `a` takes less or the same number of bytes as `ulong`. And if it's not the case, then this code compares the remaining tail of `a`'s with the remaining tail of `b`'s in bytes representation. As documentation says:
> A type is an unmanaged type if it's any of the following types:
>    * sbyte, byte, short, ushort, int, uint, long, ulong, nint, nuint, char, float, double, decimal, or bool
>    * Any [enum](https://learn.microsoft.com/en-us/dotnet/csharp/language-reference/builtin-types/enum) type
>    * Any [pointer](https://learn.microsoft.com/en-us/dotnet/csharp/language-reference/unsafe-code#pointer-types) type
>    * A [tuple](https://learn.microsoft.com/en-us/dotnet/csharp/language-reference/builtin-types/value-tuples) whose members are all of an unmanaged type
>    * Any user-defined [struct](https://learn.microsoft.com/en-us/dotnet/csharp/language-reference/builtin-types/struct) type that contains fields of unmanaged types only.

So technically it's possible that both `a` and `b` take more than `ulong`-size bytes (e.g. a tuple or some user-defined struct consisting of several fields).

> Second, this method is checking two same-type inputs, so it skips the length check.

Agree, so my question isn't relevant anymore. And given this fact the fix presented in the PR is valid.

> I was thinking about how things like this passed tests, 

I think that somehow it just happened than no one tried to call this helper for a large tuple or struct. Whereas all int-like cases are handled correctly by this helper, thanks to the first part of this method.

Though it would be good to add a unit-test that calls this method not only for int-ilke types, but also for big unmanaged tuple or struct.

File: src/Neo.Cryptography.BLS12_381/ConstantTimeUtility.cs (hunk 1 above)
Comment (AnnaShaleva) [reply src/Neo.Cryptography.BLS12_381/ConstantTimeUtility.cs:29]: We actually don't, thanks to Shili I've read a bit of documentation :D

**Comment by AnnaShaleva** [comment]
> It's true in case if a's length in serialized representation is even to ulong size.

OK, I'm wrong about it, and @txhsl is right, `a_u64.Length * sizeof(ulong)` always equals to `a_bytes.Length`. So it's just a piece of dead code. Ref. [Cast doc](https://learn.microsoft.com/en-us/dotnet/api/system.runtime.interopservices.memorymarshal.cast?view=net-8.0#system-runtime-interopservices-memorymarshal-cast-2(system-span((-0)))):

> Remarks
>
>Neither TFrom nor TTo can contain managed object references. The Cast method performs this check at runtime and throws ArgumentException if the check fails.
>
>If the sizes of the two types are different, the cast combines or splits values, which leads to a change in length.
>
>For example, if TFrom is [Int64](https://learn.microsoft.com/en-us/dotnet/api/system.int64?view=net-8.0), the ReadOnlySpan<Int64> contains a single value, 0x0100001111110F0F, and TTo is [Int32](https://learn.microsoft.com/en-us/dotnet/api/system.int32?view=net-8.0), the resulting ReadOnlySpan<Int32> contains two values. The values are 0x11110F0F and 0x01000011 on a little-endian architecture, such as x86. On a big-endian architecture, the order of the two values is reversed, i.e. 0x01000011, followed by 0x11110F0F.
>
>As another example, if TFrom is [Int32](https://learn.microsoft.com/en-us/dotnet/api/system.int32?view=net-8.0), the ReadOnlySpan<Int32> contains the values of 1, 2 and 3, and TTo is [Int64](https://learn.microsoft.com/en-us/dotnet/api/system.int64?view=net-8.0), the resulting ReadOnlySpan<Int64> contains a single value: 0x0000000200000001 on a little-endian architecture and 0x0000000100000002 on a big-endian architecture.

**Comment by cschuchardt88** [comment]
This PR patch is correct. `unmanaged` `keyword` allows upto `2GB~` of data to be passed.

**Example**:
```csharp
a_bytes = // is a struct of ulong (Like UInt160, UInt256)
b_bytes = // is a struct of ulong (Like UInt160, UInt256)
a_u64 =  // is an array of ulong
b_u64 = // is an array of ulong
```

**Review by cschuchardt88 — APPROVED** [review]

**Pull request merged**